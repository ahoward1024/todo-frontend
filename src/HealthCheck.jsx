import React from 'react';

const URL = 'http://localhost:5000/healthcheck';
const successMessage = '‚úîÔ∏è The server is up! üëå';

class HealthCheck extends React.Component {
  // We need to cancel the asyncronous task if the component is being
  // unmounted otherwise this is a memory leak.
  // Warning: Can't call setState (or forceUpdate) on an unmounted component.
  // This is a no-op, but it indicates a memory leak in your application.
  // To fix, cancel all subscriptions and asynchronous tasks in the
  // componentWillUnmount method.
  // https://stackoverflow.com/questions/50428842/cant-call-setstate-or-forceupdate-on-an-unmounted-component
  isCancelled = false;

  constructor(props) {
    super(props);
    this.state = {'message': 'Checking...'};
    this.getResponse = this.getResponse.bind(this);
  }

  async getResponse() {
    let message = 'Checking...';
    try {
      const response = await fetch(URL, {'method': 'GET'});
      const status = await response.status;
      if (!this.isCancelled && status === 200) {
        message = successMessage;
      }
    } catch (exception) {
      console.log(`Health Check Fail: ${exception}`);
      message = 'Error: '.concat(exception.toString());
    }
    if (!this.isCancelled) {
      this.setState({message});
    }
  }

  componentDidMount() {
    this.getResponse();
  }

  componentWillUnmount() {
    this.isCancelled = true;
  }

  render() {
    const style = {'color': 'white'};
    if (this.state.message === successMessage) {
      style.color = 'green';
    } else if (this.state.message.includes('Error')) {
      style.color = 'red';
    }

    return (
      <div align="center">
        <h1 className="H1-Animation">Health Check</h1>
        <p
          className="Server-Message-Animation"
          style={style}
        >{this.state.message}</p>
      </div>
    );
  }
}

export default HealthCheck;
