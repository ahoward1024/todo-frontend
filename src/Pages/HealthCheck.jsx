import React from 'react';
import config from '../config';

const URL = config.TODO_URL_HEALTHCHECK;
export const successMessage = '✔️ The server is up! 👌';
export const errorMessage = '❌ Error 👎';

class HealthCheck extends React.Component {
  // We need to cancel the asyncronous task if the component is being
  // unmounted otherwise this is a memory leak.
  // Warning: Can't call setState (or forceUpdate) on an unmounted component.
  // This is a no-op, but it indicates a memory leak in your application.
  // To fix, cancel all subscriptions and asynchronous tasks in the
  // componentWillUnmount method.
  // https://stackoverflow.com/questions/50428842/cant-call-setstate-or-forceupdate-on-an-unmounted-component
  isCancelled = false;

  constructor(props) {
    super(props);
    this.state = {
      'message': 'Checking...',
      'status': ''
    };
  }

  static async getResponse() {
    const response = await fetch(URL, {'method': 'GET'});

    return response;
  }

  componentDidMount() {
    return HealthCheck.getResponse()
    .then(response => {
      let message = '';
      if (response.status === 200) {
        message = successMessage;
      } else {
        message = errorMessage;
      }
      if (!this.isCancelled) {
        this.setState({
          message,
          'status': response.status_text
        });
      }
    })
    .catch(exception => {
      console.log(exception);
      if (!this.isCancelled) {
        const error = exception.toString();
        this.setState({
          'message': errorMessage,
          'status': error
        });
      }
    });
  }

  componentWillUnmount() {
    this.isCancelled = true;
  }

  render() {
    const style = {'color': 'white'};
    if (this.state.message === successMessage) {
      style.color = 'green';
    } else if (this.state.message.includes('Error')) {
      style.color = 'red';
    }

    return (
      <div align="center">
        <h1 className="H1-Animation">Health Check</h1>
        <p
          className="Server-Message-Animation"
          style={style}
        >
          {this.state.message}
        </p>
        <p style={{'color': 'red'}}>
          {this.state.status}
        </p>
      </div>
    );
  }
}

export default HealthCheck;
