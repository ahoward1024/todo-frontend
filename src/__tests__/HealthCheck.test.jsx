import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, {shallow, mount, render} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import HealthCheck, {successMessage, errorMessage} from '../Pages/HealthCheck';

global.fetch = require('jest-fetch-mock');

Enzyme.configure({'adapter': new Adapter()});

const mockCallback = jest.fn();
let healthCheck = '';
const response = {'test': 'message'};

describe('HealthCheck testing', () => {
  beforeEach(() => {
    fetch.resetMocks();
    healthCheck = <HealthCheck />;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('Create HealthCheck component', async () => {
    fetch.mockResponse(JSON.stringify(response));
    const tree = await renderer.create(healthCheck).toJSON();
    await expect(tree).toMatchSnapshot();
  });

  test('HealthCheck success, server is up', async () => {
    fetch.mockResponse(JSON.stringify(response), {'status': 200});
    const wrapper = await shallow(healthCheck);
    await wrapper.instance().componentDidMount();
    expect(wrapper.state('message')).toBe(successMessage);
  });

  test('HealthCheck failure, server error', async () => {
    fetch.mockResponse(JSON.stringify(response), {'status': 500});
    const wrapper = await shallow(healthCheck);
    await wrapper.instance().componentDidMount();
    expect(wrapper.state('message')).toBe(errorMessage);
  });

  test('HealthCheck component cancelled no error', async () => {
    fetch.mockResponse(JSON.stringify(response), {'status': 200});
    const wrapper = await shallow(healthCheck);
    await wrapper.instance().componentWillUnmount();
    expect(wrapper.state('message')).toBe('Checking...');
  });

  test('HealthCheck failure, server is down', async () => {
    const error =
      'Error: TypeError: NetworkError when attempting to fetch resource.';
    fetch.mockReject(error);
    const wrapper = await shallow(healthCheck);
    await wrapper.instance().componentDidMount();
    expect(wrapper.state('message')).toBe(errorMessage);
    expect(wrapper.state('status')).toBe(error);
  });

  test('HealthCheck componentWillUnmount()', async () => {
    const wrapper = await shallow(healthCheck);
    await wrapper.instance().componentWillUnmount();
    expect(wrapper.instance().isCancelled).toBe(true);
  });
});
