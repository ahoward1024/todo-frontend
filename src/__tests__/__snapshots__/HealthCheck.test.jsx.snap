// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HealthCheck testing Create HealthCheck component 1`] = `
<div
  align="center"
>
  <h1
    className="H1-Animation"
  >
    Health Check
  </h1>
  <p
    className="Server-Message-Animation"
    style={
      Object {
        "color": "white",
      }
    }
  >
    Checking...
  </p>
  <p
    style={
      Object {
        "color": "red",
      }
    }
  >
    
  </p>
</div>
`;
